processor/stream/processor.go

```diff
@@ -189,24 +186,80 @@ func (p *Processor) readBatch(
 
 	// input events are decoded and appended to the batch
 	for i := 0; i < batchSize && !reader.IsEOF(); i++ {
-		var rawModel map[string]interface{}
-		err := reader.Read(&rawModel)
+		eventType, err := p.IdentifyEventType(reader.NDJSONStreamDecoder, response)
 		if err != nil && err != io.EOF {
+			err = reader.wrapError(err)
 			if e, ok := err.(*Error); ok && (e.Type == InvalidInputErrType || e.Type == InputTooLargeErrType) {
 				response.LimitedAdd(e)
 				continue
+			} else {
+				// return early, we assume we can only recover from a input error types
+				response.Add(err)
+				return true
 			}
-			// return early, we assume we can only recover from a input error types
-			response.Add(err)
-			return true
 		}
-		if len(rawModel) > 0 {
-
-			err := p.HandleRawModel(rawModel, batch, requestTime, *streamMetadata)
-			if err != nil {
+		if eventType == "" && err == io.EOF {
+			continue
+		}
+		input := modeldecoder.Input{
+			RequestTime: requestTime,
+			Metadata:    *streamMetadata,
+			Config:      p.Mconfig,
+		}
+		if decodeFn, ok := p.models[eventType]; ok {
+			var rawModel map[string]interface{}
+			err = reader.Decode(&rawModel)
+			err = reader.wrapError(err)
+			if err != nil && err != io.EOF {
+				if e, ok := err.(*Error); ok && (e.Type == InvalidInputErrType || e.Type == InputTooLargeErrType) {
+					response.LimitedAdd(e)
+					continue
+				}
+				// return early, we assume we can only recover from a input error types
+				response.Add(err)
+				return true
+			}
+			if len(rawModel) > 0 {
+				input.Raw = rawModel[eventType]
+				if err := decodeFn(input, batch); err != nil {
+					response.LimitedAdd(&Error{
+						Type:     InvalidInputErrType,
+						Message:  err.Error(),
+						Document: string(reader.LatestLine()),
+					})
+					continue
+				}
+			}
+		} else {
+			switch eventType {
+			case "error":
+				var event model.Error
+				if handleDecodeErr(p.decodeError(reader, &input, &event), reader, response) {
+					continue
+				}
+				if p.isRUM {
+					event.RUM = true
+				}
+				batch.Errors = append(batch.Errors, &event)
+			case "span":
+				var event model.Span
+				if handleDecodeErr(p.decodeSpan(reader, &input, &event), reader, response) {
+					continue
+				}
+				if p.isRUM {
+					event.RUM = true
+				}
+				batch.Spans = append(batch.Spans, &event)
+			case "transaction":
+				var event model.Transaction
+				if handleDecodeErr(p.decodeTransaction(reader, &input, &event), reader, response) {
+					continue
+				}
+				batch.Transactions = append(batch.Transactions, &event)
+			default:
 				response.LimitedAdd(&Error{
 					Type:     InvalidInputErrType,
-					Message:  err.Error(),
+					Message:  errors.Wrap(ErrUnrecognizedObject, eventType).Error(),
```

Comment:
No big deal, but it would be less of an awkward error message if ErrUnrecognizedObject were a type, with the event type as a field: "did not recognize object type: invalid-json" vs. "invalid-json: did not recognize object type"